libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -91,3 +91,128 @@ void AP_BattMonitor_Backend::update_resistance_estimate()
     // update estimated voltage without sag
     _state.voltage_resting_estimate = _state.voltage + _state.current_amps * _state.resistance;
 }
+
+float AP_BattMonitor_Backend::voltage_resting_estimate() const
+{
+    // resting voltage should always be greater than or equal to the raw voltage
+    return MAX(_state.voltage, _state.voltage_resting_estimate);
+}
+
+AP_BattMonitor::BatteryFailsafe AP_BattMonitor_Backend::update_failsafes(void)
+{
+    const uint32_t now = AP_HAL::millis();
+
+    bool low_voltage, low_capacity, critical_voltage, critical_capacity;
+    check_failsafe_types(low_voltage, low_capacity, critical_voltage, critical_capacity);
+
+    if (critical_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.critical_voltage_start_ms == 0) {
+            _state.critical_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.critical_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Critical;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.critical_voltage_start_ms = 0;
+    }
+
+    if (critical_capacity) {
+        return AP_BattMonitor::BatteryFailsafe_Critical;
+    }
+
+    if (low_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.low_voltage_start_ms == 0) {
+            _state.low_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.low_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Low;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.low_voltage_start_ms = 0;
+    }
+
+    if (low_capacity) {
+        return AP_BattMonitor::BatteryFailsafe_Low;
+    }
+
+    // if we've gotten this far then battery is ok
+    return AP_BattMonitor::BatteryFailsafe_None;
+}
+
+bool update_check(size_t buflen, char * buffer, bool condition, const char *message) {
```

Comment:
Perhaps a better name than condition here?  maybe `bool failed`

Reply:
Could just have `return condition` at the end of `update check` instead of dual return.

Reply:
It would have to be `return !failed;` Which while it works, I'm going to deem it not worth the manipulation/reading cost.